### PR TITLE
The provider side of handling logout

### PR DESCRIPTION
**Pull request by rohe**
- [x] Any changes relevant to users are recorded in the `CHANGELOG.md`.
- [ ] The documentation has been updated, if necessary.
- [x] New code is annotated.
- [x] Changes are covered by tests.
---

### CODE REVIEW

**Comment by rohe** [comment]
The remaining Codacy error is the log line we discussed earlier.
You want me to change it ?

File: src/oic/oic/provider.py [hunk 1]
```diff
@@ -2049,3 +1883,449 @@ def get_by_sub_and_(self, sub: str, key: str, val: Any) -> Optional[str]:
...
+            "key": self.sdb.get_verify_logout(uid),
+            "redirect": redirect,
+            "action": "/" + EndSessionEndpoint("").etype,
+        }
+        return Response(
+            self.template_renderer("verify_logout", context), headers=headers
+        )
+
+    def _get_uid_from_cookie(
+        self, cookie: Union[str, SimpleCookie, None]
```
Comment (tpazderka) [inline src/oic/oic/provider.py]: This is probably `Optional[Union[str, SimpleCookie]]`?

File: src/oic/oic/provider.py [hunk 2]
```diff
@@ -2049,3 +1883,449 @@ def get_by_sub_and_(self, sub: str, key: str, val: Any) -> Optional[str]:
...
+            if _iframe:
+                res["flu"] = {client_id: _iframe}
+
+        # Clean out sessions
+        self.clean_sessions([sid])
+
+        return res
+
+    def end_session_endpoint(
+        self, request: str = "", cookie: Union[str, SimpleCookie, None] = None, **kwargs
```
Comment (tpazderka) [inline src/oic/oic/provider.py]: Again, isn't this `Optional[Union[str, SimpleCookie]]`?

File: src/oic/oic/provider.py [hunk 3]
```diff
@@ -2049,3 +1883,449 @@ def get_by_sub_and_(self, sub: str, key: str, val: Any) -> Optional[str]:
...
+        """
+        Handle a RP initiated Logout request.
+
+        :param request: The logout request
+        :param cookie:
+        :param kwargs:
+        :return: Returns a dictionary with one key 'sjwt' and the value
+            being a signed JWT token with session information.
+        """
+        esr = EndSessionRequest().from_urlencoded(request)
```
Comment (tpazderka) [inline src/oic/oic/provider.py]: Please use `self.server.message_factory.get_request_type("endsession_endpoint")` for easier overrides.

File: src/oic/oic/provider.py [hunk 4]
```diff
@@ -2049,3 +1883,449 @@ def get_by_sub_and_(self, sub: str, key: str, val: Any) -> Optional[str]:
...
+                args = {"client_id": client_id}
+            else:
+                args = {}
+
+            try:
+                id_token_hint.verify(
+                    iss=self.baseurl, skew=far_away, nonce_storage_time=far_away, **args
+                )
+            except (VerificationError, NotForMe) as err:
+                logger.warning("Verification error on id_token_hint: {}".format(err))
```
Comment (tpazderka) [inline src/oic/oic/provider.py]: Yes, please fix this so it actually logs properly.

File: src/oic/oic/provider.py [hunk 5]
```diff
@@ -2049,3 +1883,449 @@ def get_by_sub_and_(self, sub: str, key: str, val: Any) -> Optional[str]:
...
+                return error_response("invalid_request", "Bad Id Token hint")
+
+            sub = id_token_hint["sub"]
+
+            if uid:
```
Comment (tpazderka) [inline src/oic/oic/provider.py]: Test for `None` should be explicit.

File: src/oic/oic/provider.py [hunk 6]
```diff
@@ -2049,3 +1883,449 @@ def get_by_sub_and_(self, sub: str, key: str, val: Any) -> Optional[str]:
...
+                else:
+                    client_id = id_token_hint["azp"]
+
+            sids = session_get(self.sdb, "sub", sub)
+
+            for sid in sids:
+                if self.sdb[sid]["client_id"] == client_id:
+                    break
+
+        if not client_id:
```
Comment (tpazderka) [inline src/oic/oic/provider.py]: Again, test for `None` should be explicit.

File: src/oic/oic/provider.py [hunk 7]
```diff
@@ -2049,3 +1883,449 @@ def get_by_sub_and_(self, sub: str, key: str, val: Any) -> Optional[str]:
...
+                msg = "Post logout redirect URI verification failed!"
+                return error_response("invalid_request", msg)
+        else:  # If only one registered use that one
+            try:
+                _ruri = self.cdb[client_id]["post_logout_redirect_uris"]
+            except KeyError:
+                msg = "Missing post_logout_redirect_uri"
+                return error_response("invalid_request", msg)
+
+            if len(_ruri) == 1:
```
Comment (tpazderka) [inline src/oic/oic/provider.py]: `redirect_uri` is not defined for case where the `len(_ruri)` is different from `1`.

This is also not covered in tests. Either there are no logout URIs registered (and the SDB doesn't raise KeyError, which might be possible) or there are multiple.

File: src/oic/oic/provider.py [hunk 8]
```diff
@@ -2049,3 +1883,449 @@ def get_by_sub_and_(self, sub: str, key: str, val: Any) -> Optional[str]:
...
+
+        return back_channel_logout_uri, sjwt
+
+    def clean_sessions(self, usids: List[str]):
+        """
+        Remove Session IDs from the session DB.
+
+        :param usids: List of session IDs
+        """
+        _sdb = self.sdb
```
Comment (tpazderka) [inline src/oic/oic/provider.py]: This assignment seem unnecessary...

File: src/oic/oic/provider.py [hunk 9]
```diff
@@ -2049,3 +1883,449 @@ def get_by_sub_and_(self, sub: str, key: str, val: Any) -> Optional[str]:
...
+
+        :param sid: Session ID
+        :return: Dictionary with logout information
+        """
+        _sdb = self.sdb
+
+        uid = _sdb.get_uid_by_sid(sid)
+        _client_sid = {}
+        # Find all the session IDs this user has gotten
+        usids = session_get(_sdb, "uid", uid)
```
Comment (tpazderka) [inline src/oic/oic/provider.py]: I think we do not need to use `sessoin_get` for provider, since using anything else than proper SDB is discouraged for two releases.

File: src/oic/oic/provider.py [hunk 10]
```diff
@@ -2049,3 +1883,449 @@ def get_by_sub_and_(self, sub: str, key: str, val: Any) -> Optional[str]:
...
+        }  # type: Dict[str, Dict[str, Union[None, str, Tuple[str,str]]]]
+
+        if "backchannel_logout_uri" in _cdb[client_id]:
+            _sub = _sso_db[sid]["sub"]
+            _spec = self.do_back_channel_logout(_cdb[client_id], _sub, sid)
+            if _spec:
+                res["blu"] = {client_id: _spec}
+        elif "frontchannel_logout_uri" in _cdb[client_id]:
+            # Construct an IFrame
+            _iframe = do_front_channel_logout_iframe(_cdb[client_id], self.name, sid)
```
Comment (tpazderka) [inline src/oic/oic/provider.py]: I would rather have this as a staticmethod of provider, keeps it together.

File: src/oic/oic/provider.py [hunk 11]
```diff
@@ -2049,3 +1883,449 @@ def get_by_sub_and_(self, sub: str, key: str, val: Any) -> Optional[str]:
...
+                _iframe = do_front_channel_logout_iframe(_cdb[_cid], _iss, _csid)
+                if _iframe:
+                    fc_iframes[_cid] = _iframe
+
+        # Clean out all sessions
+        self.clean_sessions(usids)
+
+        return {"blu": bc_logouts, "flu": fc_iframes}
+
+    def logout_from_client(self, sid: str, client_id: str) -> Dict:
```
Comment (tpazderka) [inline src/oic/oic/provider.py]: This method would benefit for more descriptive variable names... It is sort of unreadable...

File: src/oic/oic/provider.py [hunk 12]
```diff
@@ -2049,3 +1883,449 @@ def get_by_sub_and_(self, sub: str, key: str, val: Any) -> Optional[str]:
...
+
+    def do_verified_logout(
+        self, sid: str, client_id: str, alla: bool = False, **kwargs
+    ) -> Dict:
+        """
+        Perform back channel logout and prepares the information needed for front channel logout.
+
+        :param sid: Session ID
+        :param client_id:
+        :param alla:
```
Comment (tpazderka) [inline src/oic/oic/provider.py]: Have no idea what this stands for... Either make it more descritptive, or add the docstring.

File: tests/test_oic_provider_logout.py [hunk 13]
```diff
@@ -0,0 +1,831 @@
...
+AUTHN_BROKER = AuthnBroker()
+AUTHN_BROKER.add("UNDEFINED", DummyAuthn(None, "username"))
+
+# dealing with authorization
+AUTHZ = AuthzHandling()
+SYMKEY = rndstr(16)  # symmetric key used to encrypt cookie info
+USERINFO = UserInfo(USERDB)
+
+
+class DummyHTTPClient(object):
```
Comment (tpazderka) [inline tests/test_oic_provider_logout.py]: This class can be easily replaced using `responses` which allow for cleaner mocking of HTTP requests without the need for modification of the objects. I have got rid of almost all of them and I don't want to add more in new tests.

File: tests/test_oic_provider_logout.py [hunk 14]
```diff
@@ -0,0 +1,831 @@
...
+    def __init__(self, resp_status=200):
+        self.resp_status = resp_status
+
+    def http_request(self, *args, **kwargs):
+        return Response(status_code=self.resp_status)
+
+
+class TestProvider(object):
```
Comment (tpazderka) [inline tests/test_oic_provider_logout.py]: I try to avoid using pytest in favor of unittest. The are (in my opinion) better and provide handy assert functions so you don't have to write `assert` statements.

I won't enforce this, but please consider it.

File: src/oic/oic/provider.py (hunk 1 above)
Comment (rohe) [reply src/oic/oic/provider.py]: Sort of the same but yes. Optional is just a shorthand.

File: src/oic/oic/provider.py (hunk 2 above)
Comment (rohe) [reply src/oic/oic/provider.py]: OK

File: src/oic/oic/provider.py (hunk 3 above)
Comment (rohe) [reply src/oic/oic/provider.py]: OK

File: src/oic/oic/provider.py (hunk 6 above)
Comment (rohe) [reply src/oic/oic/provider.py]: OK

File: src/oic/oic/provider.py (hunk 8 above)
Comment (rohe) [reply src/oic/oic/provider.py]: It makes things go a bit faster but if you don't care about that.

File: src/oic/oic/provider.py (hunk 9 above)
Comment (rohe) [reply src/oic/oic/provider.py]: You previously marked this as a breaking change and therefor something that should be avoided. Has that changed ?
It was the whole reason for me to do all session_* functions.

File: tests/test_oic_provider_logout.py (hunk 14 above)
Comment (rohe) [reply tests/test_oic_provider_logout.py]: Haven't used unittest these last 6-7 years. Switched to py.test because it was so much better. That might not be the case anymore ?

File: src/oic/oic/provider.py (hunk 10 above)
Comment (rohe) [reply src/oic/oic/provider.py]: OK

File: src/oic/oic/provider.py (hunk 4 above)
Comment (tpazderka) [reply src/oic/oic/provider.py]: That was a reference to the Codacy issue. This needs to be changed to percent formating and lazy syntax.

File: src/oic/oic/provider.py (hunk 8 above)
Comment (tpazderka) [reply src/oic/oic/provider.py]: That would probably depend on implementation, wouldn't it?

It is probably fine either way, was just curious.

File: tests/test_oic_provider_logout.py (hunk 14 above)
Comment (tpazderka) [reply tests/test_oic_provider_logout.py]: I believe they got much better and they are in standard library so that is a plus :0

File: src/oic/oic/provider.py (hunk 11 above)
Comment (rohe) [reply src/oic/oic/provider.py]: OK

File: src/oic/oic/provider.py (hunk 9 above)
Comment (rohe) [reply src/oic/oic/provider.py]: OK

File: tests/test_oic_provider_logout.py (hunk 13 above)
Comment (rohe) [reply tests/test_oic_provider_logout.py]: Done

File: src/oic/oic/provider.py (hunk 7 above)
Comment (rohe) [reply src/oic/oic/provider.py]: Done

File: src/oic/oic/provider.py (hunk 4 above)
Comment (tpazderka) [reply src/oic/oic/provider.py]: Hm, not sure...

File: src/oic/oic/provider.py [hunk 15]
```diff
@@ -2191,6 +2197,10 @@ def end_session_endpoint(
                     redirect_uri = "%s?%s" % (_base, query_string)
                 else:
                     redirect_uri = _base
+            else:
+                raise ValueError(
+                    "Missing post_logout_redirect_uri and more then one post_logout_redirect_uris"
```
Comment (tpazderka) [inline src/oic/oic/provider.py]: Shouldn't this result in `error_response` back to the client?

File: tests/test_oic_provider_logout.py [hunk 16]
```diff
@@ -779,9 +804,9 @@ def test_do_verified_logout_just_the_one(self):
         # I know the session backend DB is a DictSessionBackend so I can use that
         _sid = list(self.provider.sdb._db.storage.keys())[0]
 
-        self.provider.httpc = DummyHTTPClient(200)
-
+        # There is no back channel logout, hence no HTTP POST and no need for responses
```
Comment (tpazderka) [inline tests/test_oic_provider_logout.py]: Then it might be a good idea to add responses anyway as they will make sure that no communication is done. This makes this behaviour fixed as the test would otherwise fail.

File: src/oic/oic/provider.py [hunk 17]
```diff
@@ -2049,3 +1882,471 @@ def get_by_sub_and_(self, sub: str, key: str, val: Any) -> Optional[str]:
...
+        One of uid and sid MUST be provided. If uid is provided sid is ignored.
+        NO changes are made to the session DB.
+        No logout is actually performed
+        :param uid: User ID
+        :param sid: Session ID
+        :return: Dictionary with logout information
+        """
+        if not uid:
+            if not sid:
+                raise ParameterError("One of uid and sid MUST be provided")
```
Comment (tpazderka) [inline src/oic/oic/provider.py]: This part is never covered in tests.

File: src/oic/oic/provider.py [hunk 18]
```diff
@@ -2049,3 +1882,471 @@ def get_by_sub_and_(self, sub: str, key: str, val: Any) -> Optional[str]:
...
+        # Front-/Backchannel logout ?
+        _cdb = self.cdb
+        _iss = self.name
+        bc_logouts = {}
+        fc_iframes = {}
+        for _cid, _csid in _client_sid.items():
+            if "backchannel_logout_uri" in _cdb[_cid]:
+                _sub = self.sdb[_csid]["sub"]
+                _spec = self.do_back_channel_logout(_cdb[_cid], _sub, _csid)
+                if _spec:
```
Comment (tpazderka) [inline src/oic/oic/provider.py]: This is missing tests for `False` case, where `do_back_channel_logout` returns empty `_spec`. Also it should probably be an explicit test for None.

File: src/oic/oic/provider.py [hunk 19]
```diff
@@ -2049,3 +1882,471 @@ def get_by_sub_and_(self, sub: str, key: str, val: Any) -> Optional[str]:
...
+                    bc_logouts[_cid] = _spec
+            elif "frontchannel_logout_uri" in _cdb[_cid]:
```
Comment (tpazderka) [inline src/oic/oic/provider.py]: Test for  missing `back` and `front` logout uri is missing.

File: src/oic/oic/provider.py [hunk 20]
```diff
@@ -2049,3 +1882,471 @@ def get_by_sub_and_(self, sub: str, key: str, val: Any) -> Optional[str]:
...
+                # Construct an IFrame
+                _iframe = self.do_front_channel_logout_iframe(_cdb[_cid], _iss, _csid)
+                if _iframe:
```
Comment (tpazderka) [inline src/oic/oic/provider.py]: This is missing tests for False case, where do_front_channel_logout returns empty _iframe. Also it should probably be an explicit test for None.

File: src/oic/oic/provider.py [hunk 21]
```diff
@@ -2049,3 +1882,471 @@ def get_by_sub_and_(self, sub: str, key: str, val: Any) -> Optional[str]:
...
+            sids = session_get(self.sdb, "sub", sub)
+
+            for sid in sids:
+                if self.sdb[sid]["client_id"] == client_id:
+                    break
+
+        if not client_id:
+            return error_response("invalid_request", "Could not find client ID")
+        if client_id not in self.cdb:
+            return error_response("invalid_request", "Unknown client")
```
Comment (tpazderka) [inline src/oic/oic/provider.py]: This is not covered in tests.

File: src/oic/oic/provider.py [hunk 22]
```diff
@@ -2049,3 +1882,471 @@ def get_by_sub_and_(self, sub: str, key: str, val: Any) -> Optional[str]:
...
+                if self.sdb.get_uid_by_sub(sub) != uid:
+                    return error_response("invalid_request", "Wrong user")
+            else:
+                uid = self.sdb.get_uid_by_sub(sub)
+
+            if client_id is None:
+                if len(id_token_hint["aud"]) == 1:
+                    client_id = id_token_hint["aud"][0]
+                else:
+                    client_id = id_token_hint["azp"]
```
Comment (tpazderka) [inline src/oic/oic/provider.py]: This branch is not covered in tests.

File: src/oic/oic/provider.py [hunk 23]
```diff
@@ -2049,3 +1882,471 @@ def get_by_sub_and_(self, sub: str, key: str, val: Any) -> Optional[str]:
...
+
+            sids = session_get(self.sdb, "sub", sub)
+
+            for sid in sids:
```
Comment (tpazderka) [inline src/oic/oic/provider.py]: What happens if no match is found? Also that case is not covered in tests.

File: src/oic/oic/provider.py [hunk 24]
```diff
@@ -2049,3 +1882,471 @@ def get_by_sub_and_(self, sub: str, key: str, val: Any) -> Optional[str]:
...
+            logout_spec = self.logout_info_for_one_client(
+                session_id=sid, client_id=client_id
+            )
+            sids = [sid]
+
+        if self.events:
+            self.events.store("object args", "{}".format(logout_spec))
+
+        if not logout_spec["back_channel"] and not logout_spec["front_channel"]:
+            return {}
```
Comment (tpazderka) [inline src/oic/oic/provider.py]: This is missing a test.

File: src/oic/oic/provider.py [hunk 25]
```diff
@@ -2049,3 +1882,471 @@ def get_by_sub_and_(self, sub: str, key: str, val: Any) -> Optional[str]:
...
+            failed = []
+            for _cid, spec in logout_spec["back_channel"].items():
+                _url, sjwt = spec
+                logger.info("logging out from {} at {}".format(_cid, _url))
+
+                try:
+                    res = self.httpc.http_request(
+                        _url, "POST", data="logout_token={}".format(sjwt)
+                    )
+                except Exception as err:
```
Comment (tpazderka) [inline src/oic/oic/provider.py]: This is not covered in tests.

File: src/oic/oic/provider.py [hunk 26]
```diff
@@ -2049,3 +1882,471 @@ def get_by_sub_and_(self, sub: str, key: str, val: Any) -> Optional[str]:
...
+                elif res.status_code >= 400:
+                    _errstr = "failed to logout from {}".format(_cid)
+                    if self.events:
+                        self.events.store("fault", _errstr)
+                    logger.error(_errstr)
+                    failed.append(_cid)
+            # If no back-channel logout worked and there is no front-channel logout
+            # regard this as a failure.
+            if len(failed) == len(logout_spec["back_channel"]):
+                if not logout_spec["front_channel"]:
```
Comment (tpazderka) [inline src/oic/oic/provider.py]: The empty else is not covered in tests.

File: src/oic/oic/provider.py [hunk 27]
```diff
@@ -2049,3 +1882,471 @@ def get_by_sub_and_(self, sub: str, key: str, val: Any) -> Optional[str]:
...
+                    # Can't be more specific because I don't know which http client are used
+                    logger.error("failed to logout from {}".format(_cid))
+                    if self.events:
+                        self.events.store("exception", "{}: {}".format(_cid, str(err)))
+                    failed.append(_cid)
+                    continue
+
+                if res.status_code < 300:
+                    logger.info("Logged out from {}".format(_cid))
+                elif res.status_code >= 400:
```
Comment (tpazderka) [inline src/oic/oic/provider.py]: It looks to me that this can be a simple `else` since redirects should be followed, aren't they?

File: src/oic/oic/provider.py (hunk 15 above)
Comment (rohe) [reply src/oic/oic/provider.py]: Probably

File: tests/test_oic_provider_logout.py (hunk 16 above)
Comment (rohe) [reply tests/test_oic_provider_logout.py]: OK

File: src/oic/oic/provider.py (hunk 17 above)
Comment (rohe) [reply src/oic/oic/provider.py]: OK

File: src/oic/oic/provider.py (hunk 18 above)
Comment (rohe) [reply src/oic/oic/provider.py]: Well, do_back_channel_logout() is only called from the logout_for_* methods if there is a backchannel_logout_uri. So the test is not needed at all.

File: src/oic/oic/provider.py (hunk 19 above)
Comment (rohe) [reply src/oic/oic/provider.py]: Same here. Tests not needed.

File: src/oic/oic/provider.py (hunk 18 above)
Comment (tpazderka) [reply src/oic/oic/provider.py]: Then the `_spec` should never be `None` and the `if _spec` clause is not needed and should be removed.

File: src/oic/oic/provider.py (hunk 20 above)
Comment (tpazderka) [reply src/oic/oic/provider.py]: Same as above, when the condition can never be False, it shouldn't be here.

File: src/oic/oic/provider.py (hunk 19 above)
Comment (tpazderka) [reply src/oic/oic/provider.py]: No, this method can be called on logins to clients with no logout mechanics registered and such case should be covered in tests.

File: src/oic/oic/provider.py (hunk 22 above)
Comment (rohe) [reply src/oic/oic/provider.py]: OK

File: src/oic/oic/provider.py (hunk 23 above)
Comment (rohe) [reply src/oic/oic/provider.py]: OK

File: src/oic/oic/provider.py (hunk 24 above)
Comment (rohe) [reply src/oic/oic/provider.py]: OK

File: src/oic/oic/provider.py (hunk 25 above)
Comment (rohe) [reply src/oic/oic/provider.py]: OK

File: src/oic/oic/provider.py (hunk 26 above)
Comment (rohe) [reply src/oic/oic/provider.py]: Elaborate please !

Comment (tpazderka) [reply src/oic/oic/provider.py]: The case when this condition is not met is never tested.

File: src/oic/oic/provider.py (hunk 27 above)
Comment (rohe) [reply src/oic/oic/provider.py]: To assume that the client automatically follows a 30X is probably fair.

File: tests/test_oic_provider_logout.py [hunk 28]
```diff
@@ -0,0 +1,978 @@
...
+        self.provider.cdb["client0"]["client_id"] = "client0"
+
+        # Get a session ID, anyone will do.
+        # I know the session backend DB is a DictSessionBackend so I can use that
+        _sid = list(self.provider.sdb._db.storage.keys())[0]
+
+        # There is no back channel logout, hence there should be no HTTP POST
+        exception = ConnectionError()
+        with responses.RequestsMock(assert_all_requests_are_fired=False) as rsps:
+            rsps.add(responses.POST, "https://example.com/bc_logout", body=exception)
```
[the remark this answers is not in the document]
Comment (rohe) [reply tests/test_oic_provider_logout.py]: OK

File: src/oic/oic/provider.py (hunk 18 above)
Comment (rohe) [reply src/oic/oic/provider.py]: It's been removed

File: tests/test_oic_provider_logout.py [hunk 29]
```diff
@@ -0,0 +1,977 @@
...
+        "verified": True,
+        "sub": "username",
+        "extra_claim": "extra_claim_value",
+    },
+}
+
+URLMAP = {CLIENT_ID: ["https://example.com/authz"]}
+
+
+def dummy_template_renderer(template_name: str, context: Dict[str, str]) -> str:
```
Comment (tpazderka) [inline tests/test_oic_provider_logout.py]: This should probably not be here. The default template rendered can be used. And if any modifications for proper logout functionality are needed, they should be implemented (and documented) there.

File: tests/test_oic_provider_logout.py [hunk 30]
```diff
@@ -0,0 +1,1012 @@
...
+            "backchannel_logout_uri"
+        ] = "https://example.com/bc_logout"
+        self.provider.cdb["client0"]["client_id"] = "client0"
+
+        # Get a session ID, anyone will do.
+        # I know the session backend DB is a DictSessionBackend so I can use that
+        _sid = list(self.provider.sdb._db.storage.keys())[0]
+
+        # There is no back channel logout, hence there should be no HTTP POST
+        exception = ConnectionError()
```
Comment (tpazderka) [inline tests/test_oic_provider_logout.py]: The exception is now unused, but quality will probably complain about that as well :)

File: tests/test_oic_provider_logout.py [hunk 31]
```diff
@@ -0,0 +1,1012 @@
...
+            if c in self.provider.cdb["number5"]:
+                del self.provider.cdb["number5"][c]
+
+        resp = self.provider.do_verified_logout(
+            sid=list(self.provider.sdb._db.storage.keys())[0], client_id="number5"
+        )
+
+        assert resp == {}
+
+    def test_back_channel_http_exception(self):
```
Comment (tpazderka) [inline tests/test_oic_provider_logout.py]: The test name is misleading, it should probably be `test_no_back_channel`

File: src/oic/oic/provider.py (hunk 19 above)
Comment (tpazderka) [reply src/oic/oic/provider.py]: This test for `logout_info_for_all_clients` with one of the clients missing both logout specs is still missing.

Comment (rohe) [reply src/oic/oic/provider.py]: OK

File: src/oic/oic/provider.py (hunk 26 above)
Comment (rohe) [reply src/oic/oic/provider.py]: Done

File: tests/test_oic_provider_logout.py (hunk 30 above)
Comment (rohe) [reply tests/test_oic_provider_logout.py]: Removed

File: tests/test_oic_provider_logout.py (hunk 31 above)
Comment (rohe) [reply tests/test_oic_provider_logout.py]: I propose test_back_channel_logout_fails

Comment (tpazderka) [reply tests/test_oic_provider_logout.py]: Sounds good.

File: tests/test_oic_provider_logout.py (hunk 29 above)
Comment (rohe) [reply tests/test_oic_provider_logout.py]: OK

File: tests/test_oic_provider_logout.py [hunk 32]
```diff
@@ -0,0 +1,1043 @@
...
+        ] = "https://example.com/bc_logout"
+        self.provider.cdb["client0"]["client_id"] = "client0"
+
+        # Get a session ID, anyone will do.
+        # I know the session backend DB is a DictSessionBackend so I can use that
+        _sid = list(self.provider.sdb._db.storage.keys())[0]
+
+        # There is no back channel logout, hence there should be no HTTP POST
+        with responses.RequestsMock() as rsps:
+            rsps.calls  # This is just to quite pylama
```
Comment (tpazderka) [inline tests/test_oic_provider_logout.py]: The simple `with responses.RequestsMock():` shoudl do the trick as well?

Comment (rohe) [reply tests/test_oic_provider_logout.py]: No, I tried that and got complains.

Comment (tpazderka) [reply tests/test_oic_provider_logout.py]: You need to remove the assignment of the decorator, then it works (just tested it).

**Comment by tpazderka** [comment]
OK, I will squash it a little bit and merge.
